File: src/Library/Calculator/Differentiator.cpp
#include "Differentiator.h"
#include "AdditiveTerm.h"
#include "MultiplicativeTerm.h"
#include "Simplify.h"
#include "BOperNode.h"
#include "DiffOperNode.h"
#include "FValueNode.h"
#include "IntegralOperNode.h"
#include "IValueNode.h"
#include "SinusoidalOperNode.h"
#include "TensorNode.h"
#include "NOperNode.h"
#include "UOperNode.h"
#include "VarValueNode.h"

#include "Helper.h"

unique_ptr<ExpNode> Differentiator::Execute(const BOperNode& node) {
	if (node.GetChild(0).IsConstantBy(*variable_)) {
		return MakeValueNode(0);
	}

	// Multiply by exponent, decrement exponent
	const auto coefficient = node.GetChild(1);
	const auto base = node.GetChild(0);
	const auto exponent = *Add(node.GetChild(1), -1);
	return Multiply(coefficient, *Power(base, exponent));
}

unique_ptr<ExpNode> Differentiator::Execute(const DiffOperNode& node) {
	// Should not hit
	throw;
	return node.Clone();
}

unique_ptr<ExpNode> Differentiator::Execute(const FValueNode& node) {
	return MakeValueNode(0);
}

unique_ptr<ExpNode> Differentiator::Execute(const IntegralOperNode& node) {
	// Should not hit
	throw;
	return node.Clone();
}

unique_ptr<ExpNode> Differentiator::Execute(const IValueNode& node) {
	return MakeValueNode(0);
}

unique_ptr<ExpNode> Differentiator::Execute(const NOperNode& node) {
	switch (node.GetOperator())
	{
	case Operator::ADDITION:
		return ApplySumRule(node);
	case Operator::MULTIPLICATION:
		return ApplyProductRule(node);
	}

	return node.Clone();
}

unique_ptr<ExpNode> Differentiator::Execute(const SinusoidalOperNode& node) {
	// Apply ChainRule
	auto coefficient = *node.GetChild(0).Execute(this);
	// Apply derivative table
	auto sinFunc = *ApplySinusoidalTable(node);
	return Multiply(coefficient, sinFunc);
}

unique_ptr<ExpNode> Differentiator::Execute(const TensorNode& node) {
	// TODO: Tensor calculus
	return node.Clone();
}

unique_ptr<ExpNode> Differentiator::Execute(const UOperNode& node) {
	// TODO: UOper differentiation
	return node.Clone();
}

unique_ptr<ExpNode> Differentiator::Execute(const VarValueNode& node) {
	if (node.GetCharacter() != variable_->GetCharacter()) {
		return MakeValueNode(0);
	}
	else {
		return MakeValueNode(1);
	}
}

unique_ptr<ExpNode> Differentiator::ApplySumRule(const NOperNode& node) {
	unique_ptr<NOperNode> nNode = make_unique<NOperNode>(node.GetOperator());
	for (int i = 0; i < node.ChildCount(); i++)
	{
		nNode->AddChild(node.GetChild(i).Execute(this));
	}
	return move(nNode);
}

unique_ptr<ExpNode> Differentiator::ApplyProductRule(const NOperNode& node) {
	unique_ptr<NOperNode> aNode = make_unique<NOperNode>(Operator::ADDITION);
	for (int i = 0; i < node.ChildCount(); i++) {
		unique_ptr<NOperNode> mNode = make_unique<NOperNode>(Operator::MULTIPLICATION);
		mNode->AddChild(node.GetChild(i).Execute(this));
		for (int j = 0; j < node.ChildCount(); j++) {
			if (j != i) {
				mNode->AddChild(node.GetChild(j).Clone());
			}
		}
		aNode->AddChild(move(mNode));
	}
	return aNode;
}

unique_ptr<ExpNode> Differentiator::ApplySinusoidalTable(const SinusoidalOperNode& node) {
	unique_ptr<OperNode> newNode;

	switch (node.GetOperator())
	{
	case Operator::SINE:
		newNode = make_unique<SinusoidalOperNode>(Operator::COSINE);
		newNode->AddChild(node.GetChild(0).Clone());
		return newNode;
	case Operator::COSINE:
		newNode = make_unique<SinusoidalOperNode>(Operator::SINE);
		newNode->AddChild(node.GetChild(0).Clone());
		return Negative(move(newNode));
	default:
		return node.Clone();
	}
}

File: src/Library/Calculator/Operations/Differentiator.cs
// Adam Dernis © 2021

using Calculator.ExpressionTree.Nodes;
using Calculator.ExpressionTree.Nodes.Collections;
using Calculator.ExpressionTree.Nodes.Operators.BOpers;
using Calculator.ExpressionTree.Nodes.Operators.NOpers;
using Calculator.ExpressionTree.Nodes.Operators.UOpers;
using Calculator.ExpressionTree.Nodes.Operators.UOpers.SineNode;
using Calculator.ExpressionTree.Nodes.Values;
using Calculator.Operations.Abstract;

namespace Calculator.Operations
{
    /// <summary>
    /// An <see cref="Operation"/> that takes the derivate of <see cref="ExpNode"/>s.
    /// </summary>
    /// <remarks>
    /// Currently only partial derivatives.
    /// </remarks>
    public class Differentiator : Operation
    {
        private readonly VarValueNode _variable;

        /// <summary>
        /// Initializes a new instance of the <see cref="Differentiator"/> class.
        /// </summary>
        /// <param name="variable">The variable to take the derivative over.</param>
        public Differentiator(VarValueNode variable)
        {
            _variable = variable;
        }

        /// <inheritdoc/>
        public override ExpNode Execute(AdditionOperNode node)
        {
            // Sum rule
            for (int i = 0; i < node.ChildCount; i++)
            {
                ExpNode diffChild = node.GetChild(i).Execute(this);
                node.ReplaceChild(diffChild, i);
            }

            return node;
        }

        /// <inheritdoc/>
        public override ExpNode Execute(ExpNode node) => node;

        /// <inheritdoc/>
        public override ExpNode Execute(MultiplicationOperNode node)
        {
            // Product rule
            AdditionOperNode aNode = new();
            for (int i = 0; i < node.ChildCount; i++)
            {
                MultiplicationOperNode mNode = new();
                mNode.AddChild(node.GetChild(i).Clone().Execute(this));
                for (int j = 0; j < node.ChildCount; j++)
                {
                    if (j != i) mNode.AddChild(node.GetChild(j).Clone());
                }

                aNode.AddChild(mNode);
            }
            return aNode;
        }

        /// <inheritdoc/>
        public override ExpNode Execute(NumericalValueNode node)
        {
            return Helpers.MakeNumericalNode(0);
        }

        /// <inheritdoc/>
        public override ExpNode Execute(PowOperNode node)
        {
            // TODO: Handle variable in exponent
            if (node.IsConstantBy(_variable)) return Helpers.MakeNumericalNode(0);

            var coefficient = node.RightChild;
            var @base = node.LeftChild;
            var exponent = Helpers.Add(-1, coefficient);
            return Helpers.Multiply(coefficient, Helpers.Pow(@base, exponent));
        }

        /// <inheritdoc/>
        public override ExpNode Execute(ParenthesisOperNode node)
        {
            node.Child = node.Child.Execute(this);
            return node;
        }

        /// <inheritdoc/>
        public override ExpNode Execute(SineOperNode node)
        {
            if (node.IsConstantBy(_variable)) return Helpers.MakeNumericalNode(0);

            // Apply chain rule
            var coefficient = node.Child.Clone().Execute(this);
            // Apply table
            var sinFunc = SineTable(node);
            return Helpers.Multiply(coefficient, sinFunc);
        }

        /// <inheritdoc/>
        public override ExpNode Execute(TensorNode node)
        {
            for (int i = 0; i < node.ChildCount; i++)
            {
                ExpNode diffChild = node.GetChild(i).Execute(this);
                node.ReplaceChild(diffChild, i);
            }

            return node;
        }

        /// <inheritdoc/>
        public override ExpNode Execute(VarValueNode node)
        {
            return Helpers.MakeNumericalNode(node.Character == _variable.Character ? 1 : 0);
        }

        private static ExpNode SineTable(SineOperNode node)
        {
            return node.SineFunction switch
            {
                SineFunction.SINE => new SineOperNode(SineFunction.COSINE) { Child = node.Child },
                SineFunction.COSINE => Helpers.Negative(new SineOperNode(SineFunction.SINE) { Child = node.Child }),
                _ => node,
            };
        }
    }
}